### PR TITLE
Adding redux

### PR DESCRIPTION
**Pull request by mibrah42**
## What are we adding in this PR

This PR adds Redux to our application to manage the global state. Here is the directory structure for our Redux store.

```
redux/
├── store.ts
├── reducer.ts
├── actions.ts
└── types
    └── actions.ts
```

Now we're going to go through a rough breakdown for each file in the directory. The first file to consider is `store.ts`. This file exports the store instance that will be used throughout the lifetime of the application. Here are the contents of the file. 

```Typescript
// store.ts
import { createStore } from 'redux';
import { reducer } from './reducer.js';

export const store = createStore(reducer);
```

The store initializer `createStore` takes in a reducer as an argument. This reducer is essentially a function that will be called whenever a state change occurs via an 'action'. The reducer function is defined in the reducer file which is shown below.

```Typescript
// reducer.ts

...
const INITIAL_STATE: AppCreatorStore = {
  draggingWidget: null,
  editingWidget: null,
  selectedTab: Tab.widgets,
  elementAdded: false,
  reordering: false,
  widgetIDs: {
    label: 0,
    button: 0,
    select: 0,
    textbox: 0,
    panel: 0,
    slider: 0,
    checkbox: 0,
  },
};

export const reducer: Reducer<AppCreatorStore, AppCreatorAction> = (
  state = INITIAL_STATE,
  action
): AppCreatorStore => {
  switch (action.type) {
    case SET_DRAGGING_WIDGET:
      return {
        ...state,
        draggingWidget: action.payload.widget,
      };
    case SET_EDITING_WIDGET:
      return {
        ...state,
        editingWidget: action.payload.widget,
        selectedTab:
          action.payload.widget == null
            ? state.selectedTab
            : action.payload.index,
      };
    case SET_SELECTED_TAB:
      return {
        ...state,
        selectedTab: action.payload.index,
      };
    case SET_ELEMENT_ADDED:
      return {
        ...state,
        elementAdded: action.payload.value,
      };
    case SET_REORDERING:
      return {
        ...state,
        reordering: action.payload.value,
      };
    case INCREMENT_WIDGET_ID:
      return {
        ...state,
        widgetIDs: {
          ...state.widgetIDs,
          [action.payload.id]: state.widgetIDs[action.payload.id] + 1,
        },
      };
    case RESET_DRAGGING_VALUES:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};
```

The first call of the reducer function returns the initial state of the application which is declared by the `INITIAL_STATE` object. Subsequent calls occur whenever an action is dispatched. These calls return a new copy of the state with the modified values. It is important to note that Redux state is immutable and each time a state change occurs, the whole store is replaced with a new copy.

The last two files in the directory correspond to the different actions that could be taken. One file defines 'action creators' for each action, and the other file defines the interface type for each action. As an example, we are going to look at the action creator and the corresponding interface for the `SetDraggingWidgetAction` which sets the currently dragged element in a global state. 

```Typescript
// actions.ts
...
export const setDraggingWidget = (
  widget: Element | null
): SetDraggingWidgetAction => {
  return {
    type: SET_DRAGGING_WIDGET,
    payload: {
      widget,
    },
  };
};
...
```

```Typescript
// types/actions.ts
...
export interface SetDraggingWidgetAction {
  type: typeof SET_DRAGGING_WIDGET;
  payload: {
    widget: Element | null;
  };
}
...
```

Now that the setup for our Redux store is complete. We can connect our components to it and subscribe to state changes. To achieve this, we use a `connect` function that is provided for us by the `pwa-helpers` library. This will allow us to access the new state by calling a `stateChanged(state)` method whenever the state updates. In this method, we can set instance variables that will cause the component to re-render. 

As an example, we will look at how the `ActionsPanel` connects to the Redux store to re-render whenever the `selectedTab` state changes. We will also look at how events are dispatched whenever a new tab is selected.

```Typescript
// actions-panel.ts
...
import { connect } from 'pwa-helpers';
import { LitElement, customElement } from 'lit-element';

@customElement('actions-panel')
export class ActionsPanel extends connect(store)(LitElement) {
    stateChanged(state: AppCreatorStore) {
          this.selectedTab = state.selectedTab;
    }

    handleTabSwitch(index: Tab) {
          store.dispatch(setSelectedTab(index));
    }
    
    ...
    render() {
        ...
    }
}
```

As you can see, we are able to dispatch an action using `store.dispatch(...)` and passing in one of our action creators created in the actions.ts file. 

Another helpful method to note is that we have access to the state at anytime using the `store.getState()` method. This allows us to access the current state from any component without having to call the `stateChanged` method. In fact, we only need to use the `connect` function on the class declaration for those components that need to re-render on state changes. Other components that don't need to re-render, can still access the store with the getter method `store.getState()`.

We can also look into splitting the reducer into multiple smaller reducers to make it more manageable. We can also split the actions into separate files (ie. one for each action or a logical grouping).

## Alternative approach

Another approach we considered was having our own state management solution. We created a store class that keeps track of the global state throughout the application lifetime. The state was mutable and could be updated by any component in our application. We also created an event emitter which allowed components to dispatch and subscribe to events. This allowed us to re-render components on event changes.

This solution is light weight and does not require additional dependencies. Additionally, It provides enough functionality for the basic needs of our application. Nonetheless, we predicted that their could be race conditions that could occur due to its mutable nature and the lack of atomic operations. We also thought it would be better to have a well known and trusted solution that other developers are familiar with.

```Typescript
class Store {
  draggingElement: Element | null;
  editingElement: Element | null;
  elementAdded: boolean;
  reordering: boolean;
  widgetIDs: { [key: string]: number };
  events: { [key: string]: [Object, Function][] };

  constructor() {
    this.draggingElement = null;
    this.editingElement = null;
    this.elementAdded = false;
    this.reordering = false;
    this.events = {};
    this.widgetIDs = {
      label: 0,
      button: 0,
      select: 0,
      textbox: 0,
      panel: 0,
      slider: 0,
      checkbox: 0,
    };
  }

  on(eventName: string, context: Object, callback: Function) {
    if (this.events[eventName] == null) {
      this.events[eventName] = [[context, callback]];
    } else {
      this.events[eventName].push([context, callback]);
    }
  }

  remove(eventName: string, context: Object, callback: Function) {
    if (this.events[eventName]) {
      this.events[eventName].filter(
        ([ctx, cb]) => ctx !== context && cb !== callback
      );
    }
  }

  dispatch(eventName: string, ...rest: any[]) {
    if (this.events[eventName] != null) {
      this.events[eventName].forEach(([ctx, cb]) => {
        cb.apply(ctx, rest);
      });
    }
  }

  resetDraggingValues() {
    this.draggingElement = null;
    this.elementAdded = false;
    this.reordering = false;
  }
}
```

### CODE REVIEW

File: app-creator/src/widgets/empty-notice/empty-notice.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/**
```
Comment (mibrah42) [inline app-creator/src/widgets/empty-notice/empty-notice.ts:1]: Thought it would be better to create an empty notice widget since we use it in multiple places

File: app-creator/src/redux/actions.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  SetReorderingAction,
+  SET_REORDERING,
+  IncrementWidgetAction,
+  INCREMENT_WIDGET_ID,
+} from './types/actions';
+
+/**
+ * Sets the currently dragged widget to the element being dragged, or null
+ * if there isn't any.
+ * @param widget element currently being dragged.
```
Comment (joannalcy) [inline app-creator/src/redux/actions.ts]: Param comment is only required if it provides additional information: go/tsstyle#omit-comments-that-are-redundant-with-typescript. Same for several other params below.

File: app-creator/src/redux/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      };
+    case SET_SELECTED_TAB:
+      return {
+        ...state,
+        selectedTab: action.payload.index,
+      };
+    case SET_ELEMENT_ADDED:
+      return {
+        ...state,
+        elementAdded: action.payload.value,
```
Comment (joannalcy) [inline app-creator/src/redux/reducer.ts]: If `action.payload.value` is a boolean value, should we rename `elementAdded` to `isElementAdded` just to make the variable name more descriptive? Similarly, `reordering` to `isReordering`.

File: app-creator/src/widgets/attributes-tab/attributes-tab.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ *  @fileoverview The attributes-tab widget contains the different
+ *  attributes that a user can edit for a particular element.
+ */
+import { html, customElement, css, property, LitElement } from 'lit-element';
+import { nothing } from 'lit-html';
+import '@polymer/paper-input/paper-input.js';
+import '@polymer/iron-icon/iron-icon.js';
+import '@polymer/paper-input/paper-input.js';
+import '../tab-container/tab-container';
+import '../draggable-widget/draggable-widget';
```
Comment (joannalcy) [inline app-creator/src/widgets/attributes-tab/attributes-tab.ts]: Some of these imports seem unused, is there a reason they were imported?

File: app-creator/src/widgets/draggable-widget/draggable-widget.ts [hunk 5]
```diff
@@ -103,11 +114,51 @@ export class DraggableWidget extends LitElement {
     `;
   }
 
+  /**
+   * Triggered when the edit icon is clicked. Stores a reference of the selected element in the store and
+   * displays a set of inputs for editing its attributes.
+   */
+  handleEditWidget() {
```
Comment (joannalcy) [inline app-creator/src/widgets/draggable-widget/draggable-widget.ts:127]: Could we also style/highlight the widget when it is being edited?

File: app-creator/src/widgets/draggable-widget/draggable-widget.ts [hunk 6]
```diff
@@ -103,11 +114,51 @@ export class DraggableWidget extends LitElement {
...
+    }
+
+    store.dispatch(setEditingWidget(widget));
+  }
+
+  /**
+   * Returns the widget inside the draggable wrapper.
+   * @param target draggable wrapper element.
+   */
+  extractChildWidget(target: HTMLElement) {
```
Comment (joannalcy) [inline app-creator/src/widgets/draggable-widget/draggable-widget.ts]: Could you specify the return type for this function?

File: app-creator/src/widgets/draggable-widget/draggable-widget.ts [hunk 7]
```diff
@@ -103,11 +114,51 @@ export class DraggableWidget extends LitElement {
...
+    // We want to unwrap the draggable wrapper and only reference the the inner element.
+    return target.querySelector('slot')?.assignedElements()[0];
+  }
+
   /**
    * Triggered when the trash icon is clicked. If the widget is the last in the dropzone,
-   * we display the empty placeholder and center the container's flex alignments.
+   * we display the empty empty notice and center the container's flex alignments.
```
Comment (joannalcy) [inline app-creator/src/widgets/draggable-widget/draggable-widget.ts]: Extra "empty"?

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts [hunk 8]
```diff
@@ -108,12 +102,16 @@ export class Dropzone extends LitElement {
       } else {
         container.insertBefore(widgetWrapper, nextElement);
       }
+      // set the global reordering state to true so we know that we don't increment the current widget id
```
Comment (joannalcy) [inline app-creator/src/widgets/dropzone-widget/dropzone-widget.ts]: Could we break this part of function into two functions - reordering and editing, so it's clearer which case we are handling?

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts [hunk 9]
```diff
@@ -168,15 +166,16 @@ export class Dropzone extends LitElement {
...
     const container = this.shadowRoot?.getElementById(CONTAINER_ID);
     if (container == null) {
       return null;
     }
 
     const children = Array.from(container.children).filter(
       (child) =>
-        child.id !== `${widget.id}-${store.widgetIDs[widget.id]}-wrapper`
+        child.id !==
+        `${widget.id}-${store.getState().widgetIDs[widget.id]}-wrapper`
```
Comment (joannalcy) [inline app-creator/src/widgets/dropzone-widget/dropzone-widget.ts:205]: Could you add a comment about what we are filtering? Could we achieve the same result with !child.id.includes('wrapper')?

File: app-creator/src/widgets/empty-notice/empty-notice.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+ *  @fileoverview The empty-notice widget displays an icon and a message.
+ */
+import { css, customElement, html, LitElement, property } from 'lit-element';
+import { styleMap } from 'lit-html/directives/style-map';
+import '@polymer/paper-card/paper-card.js';
```
Comment (joannalcy) [inline app-creator/src/widgets/empty-notice/empty-notice.ts]: Is this used somewhere?

**Review by joannalcy — COMMENTED** [review]
Thanks for the well-documented Redux solution! It made reviewing much easier. :)

**Comment by sufyanAbbasi** [comment]
> Thanks for the well-documented Redux solution! It made reviewing much easier. :)

+1, the PR description is comprehensive and gave me all the info I needed to feel confident about a review!

File: app-creator/src/redux/actions.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      draggingElement: null,
+      elementAdded: false,
+      reordering: false,
+    },
+  };
+};
+
+/**
+ * Sets state to true if an element has been added to a dropzone. This is used to increment
+ * a specified widget's id.
```
Comment (sufyanAbbasi) [inline app-creator/src/redux/actions.ts]: nit: Capitalize "ID" if used in a sentence (also below).

File: app-creator/src/redux/reducer.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Initial state of our application.
+ */
+const INITIAL_STATE: AppCreatorStore = {
+  draggingWidget: null,
+  editingWidget: null,
+  selectedTab: Tab.widgets,
+  elementAdded: false,
+  reordering: false,
```
Comment (sufyanAbbasi) [inline app-creator/src/redux/reducer.ts]: I was wondering if you considered using an enum to combine these events into one? This means that, half the time, components are listening to the same drag event but result in a noop based on the value of the enum. I feel that having less events in our system makes searching and following codepaths easier, since you can search for all occurrences of the event simultaneously. Happy to chat about it before committing to a decision!

Comment (sufyanAbbasi) [reply app-creator/src/redux/reducer.ts]: Same could be true about dragging and editing widget, both of these fields should be mutually exclusive, since we don't expect a user to be dragging and editing a widget at the same time. 

For example, the store can simply be:
```
{
  element: Element | null; // The main widget being interacted with.
  eventType: "dragging"|"editing"|"reordering"|"adding"
  widgetIDs: ...
}
```

File: app-creator/src/widgets/actions-panel/actions-panel.ts [hunk 13]
```diff
@@ -82,21 +92,24 @@ export class ActionsPanel extends LitElement {
     const widgetsMarkup =
       selectedTab === 1 ? html`<widgets-tab></widgets-tab>` : nothing;
 
+    const attributesMarkup =
+      selectedTab === 2 ? html`<attributes-tab></attributes-tab>` : nothing;
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/actions-panel/actions-panel.ts]: This is better expressed as a switch statement:

```
let renderedTab = nothing;
switch (selectedTab)
  case Tab.widgets:
   ...
  case Tab.templates:
   ...
}
```

Then just render `renderedTab` below.

File: app-creator/src/widgets/attributes-tab/attributes-tab.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      this.editingWidget == null
+        ? html`
+            <empty-notice
+              icon="create"
+              message="No widget selected. Click on a widget's edit icon to select it."
+              size="large"
+              bold
+            ></empty-notice>
+          `
+        : nothing;
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/attributes-tab/attributes-tab.ts]: Move the ternary logic into the return html below, to be more explicit:
```
const emptyNotice = html`
         <empty-notice
             icon="create"
              message="No widget selected. Click on a widget's edit icon to select it."
              size="large"
              bold
            ></empty-notice>
          `
...

<tab-container>${this.editingWidget ? nothing : emptyNotice}</tab-container>
```

File: app-creator/src/widgets/attributes-tab/test/attributes-tab_test.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  test('is defined', () => {
+    const el = document.createElement('attributes-tab');
+    assert.instanceOf(el, AttributesTab);
+  });
+
+  test('renders widget', async () => {
+    const el = await fixture(html`<attributes-tab></attributes-tab>`);
+    expect(el.shadowRoot!.childNodes.length).to.be.greaterThan(0);
+  });
+});
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/attributes-tab/test/attributes-tab_test.ts:22]: Add a test for rendering the empty notice when editing widget is null.

File: app-creator/src/widgets/draggable-widget/draggable-widget.ts [hunk 16]
```diff
@@ -151,25 +202,29 @@ export class DraggableWidget extends LitElement {
...
    */
   handleDragend() {
-    const draggingElement = store.draggingElement;
-    if (draggingElement && store.elementAdded && !store.reordering) {
-      store.widgetIDs[draggingElement.id]++;
+    const draggingWidget = store.getState().draggingWidget;
+    if (
+      draggingWidget &&
+      store.getState().elementAdded &&
+      !store.getState().reordering
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/draggable-widget/draggable-widget.ts]: Re: consolidating the state logic comment above.

For example, this check would become a little easier to read:
```
if (
  element &&
  store.getState().eventType === 'adding' // Better, an enum type
 )
```

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts [hunk 17]
```diff
@@ -91,8 +85,8 @@ export class Dropzone extends LitElement {
     }
 
     // Get widget that's currently being dragged.
-    const widget = store.draggingElement as HTMLElement;
-    const widgetWrapper = widget.parentElement;
+    const widget = store.getState().draggingWidget;
+    const widgetWrapper = widget?.parentElement;
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/dropzone-widget/dropzone-widget.ts]: Nice optional chaining here!

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts (hunk 9 above)
Comment (mibrah42) [reply app-creator/src/widgets/dropzone-widget/dropzone-widget.ts:205]: So here we are grabbing all the child widgets that are not the currently dragged widget. Since the direct children of the dropzone are the draggable wrappers. We are checking for the wrapper of the currently dragged widget. Therefore we need to check  `${widget.id}-${store.getState().widgetIDs[widget.id]}-wrapper`. I added a comment to clarify this. :)

Comment (joannalcy) [reply app-creator/src/widgets/dropzone-widget/dropzone-widget.ts:205]: Oh I see! I currently don't see any added comment here, maybe it's not pushed yet? But adding a comment to explain this would be great, thanks! :)

**Review by joannalcy — APPROVED** [review]

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts (hunk 9 above)
Comment (mibrah42) [reply app-creator/src/widgets/dropzone-widget/dropzone-widget.ts:205]: I just pushed an update. I'm still trying out a refactor based on @sufyanAbbasi comment:

> Same could be true about dragging and editing widget, both of these fields should be mutually exclusive, since we don't expect a user to be dragging and editing a widget at the same time.
> 
> For example, the store can simply be:
> 
> ```
> {
>   element: Element | null; // The main widget being interacted with.
>   eventType: "dragging"|"editing"|"reordering"|"adding"
>   widgetIDs: ...
> }
> ```
https://github.com/googleinterns/earthengine-community/pull/5#discussion_r433396091

File: app-creator/src/widgets/dropzone-widget/dropzone-widget.ts [hunk 18]
```diff
@@ -168,15 +188,20 @@ export class Dropzone extends LitElement {
...
    * @param y the y coordinate of the triggered event.
    */
-  getNextElement(widget: HTMLElement, y: number): Element | null {
+  getNextElement(widget: Element, y: number): Element | null {
     const container = this.shadowRoot?.getElementById(CONTAINER_ID);
     if (container == null) {
       return null;
     }
 
+    /** Get all widgets in the container excluding the currently dragged widget.
```
Comment (sufyanAbbasi) [inline app-creator/src/widgets/dropzone-widget/dropzone-widget.ts]: nit: newline.

**Review by sufyanAbbasi — APPROVED** [review]
Nice job with the refactor and I'm looking forward to the next PR!
